### PR TITLE
MONGOID-4716 Fix eager loading of nested referenced associations

### PR DESCRIPTION
**Pull request by Neilshweky**
MONGOID-4716

### CODE REVIEW

File: spec/mongoid/criteria/includable_spec.rb [hunk 1]
```diff
@@ -0,0 +1,1317 @@
...
+      let!(:documents) do
+        criteria.entries
+      end
+
+      it "returns the correct documents" do
+        expect(criteria).to eq([ person ])
+      end
+    end
+
+    context "when including nested referenced associations" do
```
Comment (Neilshweky) [inline spec/mongoid/criteria/includable_spec.rb:1080]: Here is the test that I wrote

**Comment by p-mongo** [comment]
Can you please make the first commit its own PR?

File: lib/mongoid/association/eager_loadable.rb [hunk 2]
```diff
@@ -20,13 +20,31 @@ def eager_load(docs)
...
+      def preload(associations, docs)
+        assoc_map = associations.group_by(&:inverse_class_name)
+        docs_map = { klass.to_s => docs.to_set }
+        queue = [ klass.to_s ]
+
+        while klass = queue.shift
+          if as = assoc_map.delete(klass)
+            as.group_by(&:relation)
+            .each do |relation, assocs|
+              assocs.each { |a| queue << a.class_name }
```
[the remark this answers is not in the document]
Comment (Neilshweky) [reply lib/mongoid/association/eager_loadable.rb:38]: I think I have something similar in includable_spec.rb:1242 but I will add these tests explicitly.

**Comment by Neilshweky** [comment]
@p-mongo is this PR title okay? Feel free to change it if it's not what you had in mind.

File: lib/mongoid/association/eager_loadable.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (Neilshweky) [reply lib/mongoid/association/eager_loadable.rb:38]: done

**Review by p-mongo — APPROVED** [review]

**Pull request merged**